presto-spi/src/main/java/io/prestosql/spi/type/SqlTimestamp.java

```diff
@@ -121,4 +122,16 @@ public String toString()
     {
         return formatTimestamp(precision, epochMicros, picosOfMicros);
     }
+
+    /**
+     * @return timestamp rounded to nanosecond precision
+     */
+    public LocalDateTime toLocalDateTime()
+    {
+        long epochSecond = epochMicros / MICROSECONDS_PER_SECOND;
+        int microOfSecond = toIntExact(epochMicros % MICROSECONDS_PER_SECOND);
+        int nanoOfSecond = (microOfSecond * NANOSECONDS_PER_MILLISECOND) +
+                toIntExact(round(picosOfMicros, 3) / PICOSECONDS_PER_NANOSECOND);
```

Comment:
You can use `roundDiv(picosOfMicros, PICOSECONDS_PER_NANOSECOND)` instead of `round(picosOfMicros, 3) / PICOSECONDS_PER_NANOSECOND`. Generally, I avoid using `round()` directly unless the expression has a clear indication of what the magnitude is (e.g, `12 - targetPrecision`, etc). The `3` in this case is a relative magnitude between picoseconds and nanoseconds, but it's not immediately apparent.

Reply:
Yeah, this is more clear.